### PR TITLE
qa: avoid reusing CephManager object across tests

### PR DESCRIPTION
**Pull request by mchangir**
Since the ceph cluster is destroyed and recreated after every test case, it wasn't a good idea to reuse the same CephManager object.

Introduced-by: 329edd3c56baff81cb76c5008ac2eaf7c110f9e1
Fixes: https://tracker.ceph.com/issues/66009
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by batrick — CHANGES_REQUESTED** [review]
Huh? Why would this cause the ceph-mgr `fs volume ls` command to hang?

The intent behind @leonid-s-usov 's change was to have the `Filesystem` use the `LocalCephManager`. Since it was previously instantiating its own `CephManager`, we couldn't override it. We shouldn't simply revert.

**Comment by mchangir** [comment]
jenkins test make check

**Comment by mchangir** [comment]
> Huh? Why would this cause the ceph-mgr `fs volume ls` command to hang?

`fs volume ls` command times out after 120 seconds. With Leonid's changes, the manager doesn't respond within 120 seconds and so the test is marked as failed.
When the timeout for the command is bumped up to 900 seconds, the test passes without any problems.
I'm not so sure about the **why** aspect of the command **stall**.
Probably delay in receiving/retrieving updated mds map.
Surprisingly, the `wait_for_daemons()` in `CephFSTestCase::setUp()` succeeds signalling that the mds is in `up:active` state. But later the `fs volume ls` command stalls and causes a test timeout

> 
> The intent behind @leonid-s-usov 's change was to have the `Filesystem` use the `LocalCephManager`. Since it was previously instantiating its own `CephManager`, we couldn't override it. We shouldn't simply revert.

Since the intent behind Leonid's changes wasn't mentioned in the commit message, this was the most obvious change that I could think of that solved *my* problem.

Could we have a factory of classes defined for the `vstart` and `teuthology` environments ?
We could set the environment specific factory in the startup code and then use the factory to instantiate the necessary classes.

Since the local variants of the classes are derived out of the non-local variants, having cleanly separated factory of classes seems cumbersome.

Maybe Leonid's change could go in `def setup_manager()` in `qa/tasks/ceph.py` with the necessary `if..else` kludging for local and non-local environments.

**Comment by mchangir** [comment]
The `LocalCephCluster` class in `vstart_runner.py` is creating the `LocalCephManager` object.
There's also a comment in the `LocalCephCluster::__init__()` method that the base class (`CephClusterBase`) constructor is intentionally being avoided.

This means that `LocalCephCluster` class deals with `LocalCephManager` and `CephClusterBase` class deals with `CephManager`.

I would like to urge @leonid-s-usov to give this branch a run locally for his quiescer tests and see if my changes affect his tests at all. My hypothesis is that my changes won't affect his quiescer testing on a vstart cluster.

**Comment by leonid-s-usov** [comment]
I'm sorry to see the confusion here. The `CephClusterBase` comes from a different commit than this PR addresses. Both commits are from https://github.com/ceph/ceph/pull/55823, but the one with `CephClusterBase` can't be the cause when instance reuse is the culprit. That commit, as Patrick noted, does the opposite: it allows the creation of new instances by the existing code while maintaining control over the implementation.

The same technique could be applied to the case of `CephManager`, but at the time of writing a different approach was chosen, to share the instance and inject it via the context.

If you believe that the reuse of the ceph manager instance across teuthology tasks is incorrect, then you should provide an alternative implementation to the dependency injection, so that vstart_runner would be able to provide its implementation of the manager.

**Comment by mchangir** [comment]
@leonid-s-usov Could you briefly describe what test fails for you without the bit of code you added to `CephClusterBase` to reuse the instance so that I could give it a run and make some observations and understand the situation better. I hope this isn't too much to ask for.

**Comment by mchangir** [comment]
> @mchangir I haven't run this branch. My main point was to explain the rationale behind the commit which introduced the instance sharing. I hope you test your changes by running the quiescer locally with the example provided in the commit message of [qa/tasks: vstart_runner: introduce --config-mode](https://github.com/ceph/ceph/pull/55823/commits/274849e544dd1f77158a2c80a4c654cb0363f71d)

thanks, this is good enough

File: qa/tasks/cephfs/test_snap_schedules.py [hunk 1]
```diff
@@ -41,12 +41,23 @@ def remove_snapshots(self, dir_path, sdn):
             self.mount_a.run_shell(['sudo', 'rmdir', snapshot_path])
 
     def get_snap_dir_name(self):
-        from .fuse_mount import FuseMount
-        from .kernel_mount import KernelMount
+        km = None
+        fm = None
+
+        if 'Local' in str(type(self.ceph_cluster)):
```
Comment (leonid-s-usov) [inline qa/tasks/cephfs/test_snap_schedules.py:47]: Using literal match with a class name is not a good use of Python.

File: qa/tasks/cephfs/test_snap_schedules.py [hunk 2]
```diff
@@ -41,12 +41,23 @@ def remove_snapshots(self, dir_path, sdn):
...
+            from tasks.vstart_runner import LocalFuseMount
```
Comment (leonid-s-usov) [inline qa/tasks/cephfs/test_snap_schedules.py:48]: You should not be importing vstart_runner from the core test code. that's against the whole point of dependency injection.

**Review by leonid-s-usov — CHANGES_REQUESTED** [review]
Well, even if this works - and I believe it does - it's far from what I would accept as an alternative solution.

Please consider the following, by the example of the other classes:

1. rename the class `tasks.ceph_manager.CephManager` to `CephManagerBase`
2. declare a module variable in `tasks.ceph_manager` as `CephManager = CephManagerBase`
3. in `vstart_runner.py` assign `LocalCephManager` to the new `CephManager` module var

with this, you should revert my changes in `cephfs.filesystem.__init__` back to an unconditional creation of the CephManager instance. With the change number 3 above it will be an instance of the LocalCephManager, and in other environments it will default to the CephManagerBase

**Comment by vshankar** [comment]
> Since the ceph cluster is destroyed and recreated after every test case, it wasn't a good idea to reuse the same CephManager object.

@mchangir - could you explain a bit about what's happens when reusing the same CephManager object. Is it due to a stale ceph-mgr connection hanging around that the test code uses to dispatch commands to ceph-mgr that causes this behaviour? Also, update the commit message with this information.

**Comment by mchangir** [comment]
> > Since the ceph cluster is destroyed and recreated after every test case, it wasn't a good idea to reuse the same CephManager object.
> 
> @mchangir - could you explain a bit about what's happens when reusing the same CephManager object. Is it due to a stale ceph-mgr connection hanging around that the test code uses to dispatch commands to ceph-mgr that causes this behaviour? Also, update the commit message with this information.

yeah, I'll update this PR with the latest info once I'm convinced what exactly is happening when sharing the object across tests
this is still a work-in-process issue

File: qa/tasks/ceph_test_case.py [hunk 3]
```diff
@@ -148,6 +148,10 @@ def _init_mon_manager(self):
             from tasks.ceph_manager import CephManager
```
Comment (leonid-s-usov) [inline qa/tasks/ceph_test_case.py:148]: Reviewing the older part of `_init_mon_manager` I see that you weren't the first to use `if 'Local' in str(type(self.ceph_cluster)):`. It doesn't make the code valid, but it explains the origin.

I would have this fixed. The manager instance should either be taken from the context, as I did originally, or it should implement the same class interface as `CephManager`, and should be injected via the module class variable, as I did with the other vstart_runner overrides.

No code outside of vstart_runner should be using classes defined there

**Comment by batrick** [comment]
status on this?

**Comment by vshankar** [comment]
> status on this?

I spoke to @mchangir on this today. It seems like reverting commit 329edd3c56baff81cb76c5008ac2eaf7c110f9e1 causes the tests to pass, but it isn't totally clear what the problem in the commit is. @mchangir is running some tests for explaining that.

**Comment by mchangir** [comment]
This PR has been superceded by: https://github.com/ceph/ceph/pull/58547
Closing this PR.

**Pull request closed without merging**